File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.py
class Solution:
    def smallestPalindrome(self, s: str) -> str:
        n = len(s)
        res = ['a' for _ in range(n)]
        alpha = [0 for _ in range(26)]
        i = 0
        for ch in s:
            alpha[ord(ch) - ord('a')] += 1
        for idx in range(26):
            freq = alpha[idx]
            if freq == 1:
                res[n // 2] = chr(idx + ord('a'))
            elif freq >= 2:
                for j in range(freq // 2):
                    res[i] = chr(idx + ord('a'))
                    res[-i - 1] = chr(idx + ord('a'))
                    freq -= 2
                    i += 1
                if freq == 1:
                    res[n // 2] = chr(idx + ord('a'))
        return "".join(res)

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        int n = s.size();
        string res(n, 'a');
        vector<int> alpha(26, 0);
        int idx = 0;
        for(int i = 0; i < n; i++)
        {
            alpha[s[i] - 'a']++;
        }
        for(int i = 0; i < 26; i++)
        {
            int freq = alpha[i];
            if(freq > 0)
            {
                for(int j = 0; j < alpha[i] / 2; j++)
                {
                    res[idx] = 'a' + i;
                    res[n - idx - 1] = 'a' + i;
                    freq -= 2;
                    idx++;
                }
                if(freq == 1)
                {
                    res[n / 2] = 'a' + i;
                }
            }
        }
        return res;
    }
};